### DISCUSSION TITLE
Site suggestions

### DISCUSSION THREAD

**Discussion by TudorHulban** [mint-lang/mint#462] [General]
Hi!
Some suggestions, maybe useful.
Thank you for the nice work with mint.

1. Add routing blog example to Routing section. 
2. https://www.mint-lang.com/blog/routing there is typo in prevent default link.
3. Would suggest to open external links in new tab. Currently you open them in same tab. 
This way the tab with mint site can easily remain and back is not needed (better UX). 
4. Consolidate the web sockets example in a blog / documentation entry.
5. I know in Elm Richard Feldman has in one of its courses an example with a table which he does autosearch and sorting as pinnacle (maybe) of his course. An implementation in mint as an example would be useful, anyway the object is needed for pages.
6. Is there a way to protect routes or is it not necessary?

**Comment by TudorHulban** [comment]
In VS Code is there any mint extension that would support Emmet actions in mint functions?

**Comment by s0kil** [comment]
@TudorHulban There is no such extension yet.

**Comment by s0kil** [comment]
@TudorHulban What do you mean by "protect routes" ?

**Comment by TudorHulban** [comment]
We could do some checking before redirecting to a protected route, ex.:
a. check if credentials / token still valid
b. while being authorized check if user is accessing his profile page and not someone's else

Resources: https://tylermcginnis.com/react-router-protected-routes-authentication/ .

**Comment by s0kil** [comment]
Here is an example in the [Mint Realworld](https://github.com/mint-lang/mint-realworld) project that describes what your looking for:
https://github.com/mint-lang/mint-realworld/blob/master/source/Routes.mint#L82-L96

**Comment by gdotdesign** [comment]
In Mint there is no language construct to handle this. We usually do it inside the route handler: https://github.com/mint-lang/mint-realworld/blob/master/source/Routes.mint#L82
In there the `Application.initialize()` loads the user into `Application.user` and based on that we either proceed or redirect.

**Comment by TudorHulban** [comment]
Understood. Would suggest to add this to the documentation. It is hard to the new coming user to pick up the features you offer.
Check another router to see the detail of documentation:
https://github.com/jorgegorka/svelte-router

Would windows.navigate be the best way for the documentation example? 
https://stackoverflow.com/questions/948227/should-i-use-window-navigate-or-document-location-in-javascript

**Comment by gdotdesign** [comment]
> Understood. Would suggest to add this to the documentation. It is hard to the new coming user to pick up the features you offer.

Thanks, sadly the documentation is not 100% and any input is welcome, will try to work in it in the near future.

> Would windows.navigate be the best way for the documentation example?

It's neither `window.navigate` or `window.location.href = url`. `Window.navigate` Mint function calls `window.history.pushState` under the hood: https://github.com/mint-lang/mint-runtime/blob/master/source/Utils.js#L8

And yes It would be the best way for a documentiation example.

I reopened the issue to track the documentation requests.

**Comment by TudorHulban** [comment]
Understood. `Windows.navigate` is quite similar to `windows.navigate`. 
I believe it needs to be made clear in the documentation. 

I believe it is important to expand the router section in the main site with adding clear instructions on how to perform route protection.

What are the features that you presently miss? Maybe it is a time to work also on documentation, I see features that you already developed for many months but they are not in main site documentation.

I might give it another go when you also add websockets in main site. This way the solution becomes interesting to choose. Just drop me a line ....

**Comment by Sija** [comment]
> In Mint there is no language construct to handle this. We usually do it inside the route handler: https://github.com/mint-lang/mint-realworld/blob/master/source/Routes.mint#L82
> In there the `Application.initialize()` loads the user into `Application.user` and based on that we either proceed or redirect.

@gdotdesign That's pretty verbose and not rly DRY, perhaps we could add a concept of guards, matching other routers in this regard?

**Comment by gdotdesign** [comment]
I'm not sure how we could DRY it up more in a language level without increasing the complexity a lot. I'm open to suggestions.

**Comment by TudorHulban** [comment]
Not related to your question but maybe a sugar candy? 
Window.navigate = redirect.
a. Easier to remember. 
b. No eventual association to the 'window` object.
